sdk/storage/storage-file-datalake/review/storage-file-datalake.api.md

```diff
@@ -294,6 +302,21 @@ export interface FileFlushOptions extends CommonOptions {
     retainUncommittedData?: boolean;
 }
 
+// @public
+export interface FileParallelUploadOptions extends CommonOptions {
+    abortSignal?: AbortSignalLike;
+    chunkSize?: number;
+    close?: boolean;
+    conditions?: DataLakeRequestConditions;
+    initialTransferSize?: number;
+    maxConcurrency?: number;
+    metadata?: Metadata;
+    onProgress?: (progress: TransferProgressEvent) => void;
+    pathHttpHeaders?: PathCreateHttpHeaders;
```

Comment:
Nice sugestion. Will open a separate issue for adopting `OperationOptions` but won't do it now as we are in a hurry to ship this March release.
And will rename it to `httpHeaders` for now.

Reply:
It would be a lot of work to take OperationOptions as it is, so I'm not suggesting that. However, if you stick to the same shapes as OperationOptions, that eases adoption in the future because the types will be compatible.

Reply:
But customHeaders is not quite specific about the keys?
`customHeaders?: { [key: string]: string };`

Reply:
Are you suggesting
```
export interface FileParallelUploadOptions extends CommonOptions {
...
    onProgress?: (progress: TransferProgressEvent) => void;
    httpHeaders?: { [key: string]: string };
...
}
```
We think this is not user-friendly. If not, could you please be more elaborate?

Reply:
The suggestion is:

```
export interface FileParallelUploadOptions extends CommonOptions {
...
    onProgress?: (progress: TransferProgressEvent) => void;
    requestOptions: {
        customHeaders?: PathCreateHttpHeaders;
    },
...
}
```

It's true that the `OperationOptions` interface's `requestOptions.customHeaders` object is just a type with a string indexer (`{ [key: string]: string }`), but you can safely provide a stricter type if you like. So imagine a world where `CommonOptions extends corehttp.OperationOptions` which is typed like above. In that case, we will be able to pass our more specific `CommonOptions` type (with well-defined headers) to any method in core-http that is expecting the base OperationOptions type:

```
type OperationOptionsLike = { [s: string]: string };
type CommonOptionsLike = { contentType: string };

declare function sendOperationRequest(o: OperationOptionsLike): void;
declare const opts: CommonOptionsLike;
opts.contentType // (note how we get intellisense)
sendOperationRequest(opts); // but we can still pass in our more specific type no problem
```

That said, I'm willing to believe that there are good reasons to lift these headers up to a top-level option. For example, if we believe setting these headers will almost always happen in practice, making them more accessible is good. I wish I had more time to ponder :)

Reply:
Do you happen to know what other languages are doing with this headers property?

Reply:
I see. One more question, are we settled on extending the `coreHttp.OperationOptions`. Because I find currently params in our generated code is extending `coreHttp.RequestOptionsBase` which has a less nesting for `customHeaders`.
```
export interface RequestOptionsBase {
  /**
   * @property {object} [customHeaders] User defined custom request headers that
   * will be applied before the request is sent.
   */
  customHeaders?: { [key: string]: string };

  /**
   * The signal which can be used to abort requests.
   */
  abortSignal?: AbortSignalLike;
...
```


.NET is using `PathHttpHeaders`  in the paramter, and it's flattend in the protocal (REST) layer.
Not a good reference.
```
        public virtual Response<PathInfo> Upload(
            string path,
            PathHttpHeaders httpHeaders = default,
            DataLakeRequestConditions conditions = default,
            IProgress<long> progressHandler = default,
            StorageTransferOptions transferOptions = default,
            CancellationToken cancellationToken = default) {}
```

Reply:
The new generator we're working on should generate code extending OperationOptions rather than RequestOptionsBase. OperationOptions is the new way of doing things for user-facing methods - ROB is kind of a mess and should never be exposed directly to the user.

Reply:
I talked with other architects, and I think we should go with your original approach and make no changes. It will align with what other libraries are doing, and @johanste made a good point that generally speaking we shouldn't encourage users to be setting custom headers.

Very sorry for the randomization here, and thanks for thinking this through with me!!